std/math.d

```diff
@@ -1722,159 +1993,334 @@ auto sqrt(creal z) @nogc @safe pure nothrow
+        static immutable T[5] P = [
+            9.999999999999999999999999999999999998502E-1L,
+            3.508710990737834361215404761139478627390E-2L,
+            2.708775201978218837374512615596512792224E-4L,
+            6.141506007208645008909088812338454698548E-7L,
+            3.279723985560247033712687707263393506266E-10L
+        ];
+        static immutable T[6] Q = [
+            2.000000000000000000000000000000000000150E0,
+            2.368408864814233538909747618894558968880E-1L,
+            3.611828913847589925056132680618007270344E-3L,
+            1.504792651814944826817779302637284053660E-5L,
+            1.771372078166251484503904874657985291164E-8L,
+            2.980756652081995192255342779918052538681E-12L
+        ];
+
+        // C1 + C2 = LN2.
+        enum T C1 = 6.93145751953125E-1L;
+        enum T C2 = 1.428606820309417232121458176568075500134E-6L;
+
+        // Overflow and Underflow limits.
+        enum T OF =  1.135583025911358400418251384584930671458833e4L;
+        enum T UF = -1.143276959615573793352782661133116431383730e4L;
+    }
+    else
+        static assert(0, "Not implemented for this architecture");
 
-/**
- * Calculates e$(SUPERSCRIPT x).
- *
- *  $(TABLE_SV
- *    $(TR $(TH x)             $(TH e$(SUPERSCRIPT x)) )
- *    $(TR $(TD +$(INFIN))     $(TD +$(INFIN)) )
- *    $(TR $(TD -$(INFIN))     $(TD +0.0)      )
- *    $(TR $(TD $(NAN))        $(TD $(NAN))    )
- *  )
- */
-real exp(real x) @trusted pure nothrow @nogc
-{
-    version(D_InlineAsm_X86)
+    // Special cases. Raises an overflow or underflow flag accordingly,
+    // except in the case for CTFE, where there are no hardware controls.
+    if (isNaN(x))
+        return x;
+    if (x > OF)
     {
-        //  e^^x = 2^^(LOG2E*x)
-        // (This is valid because the overflow & underflow limits for exp
-        // and exp2 are so similar).
-        return exp2(LOG2E*x);
+        if (__ctfe)
+            return T.infinity;
+        else
+            return T.max * copysign(T.max, T.infinity);
     }
-    else version(D_InlineAsm_X86_64)
+    if (x < UF)
     {
-        //  e^^x = 2^^(LOG2E*x)
-        // (This is valid because the overflow & underflow limits for exp
-        // and exp2 are so similar).
-        return exp2(LOG2E*x);
+        if (__ctfe)
+            return cast(T) 0.0;
+        else
+            return T.min_normal * copysign(T.min_normal, cast(T) 0.0);
+    }
+
+    // Express: e^^x = e^^g * 2^^n
+    //   = e^^g * e^^(n * LOG2E)
+    //   = e^^(g + n * LOG2E)
+    T xx = floor((cast(T) LOG2E) * x + cast(T) 0.5);
+    const int n = cast(int) xx;
+    x -= xx * C1;
+    x -= xx * C2;
+
+    static if (F.realFormat == RealFormat.ieeeSingle)
+    {
+        xx = x * x;
+        x = ((((( 1.9875691500E-4f  * x
+                + 1.3981999507E-3f) * x
+                + 8.3334519073E-3f) * x
+                + 4.1665795894E-2f) * x
+                + 1.6666665459E-1f) * x
+                + 5.0000001201E-1f) * xx
```

Comment:
Although. Is this still needed when you have a static foreach poly implementation?

Reply:
Not really I guess. I changed it accordingly; it's not like the D port has been a 1:1 copy of the C code anyway, so deviating some more is probably okay.

Reply:
The DMD timings for the float-versions seem to suffer by about 10% after changing to `poly()`. I've seen similar slowdowns when adding a new variable (`const T xx = x*x`) instead of reusing an existing one (`z = x*x`) with DMD too though, so... ;)